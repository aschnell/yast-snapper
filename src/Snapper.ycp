/* ------------------------------------------------------------------------------
 * Copyright (c) 2006 Novell, Inc. All Rights Reserved.
 *
 *
 * This program is free software; you can redistribute it and/or modify it under
 * the terms of version 2 of the GNU General Public License as published by the
 * Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with
 * this program; if not, contact Novell, Inc.
 *
 * To contact Novell about this file by physical or electronic mail, you may find
 * current contact information at www.novell.com.
 * ------------------------------------------------------------------------------
 */

/**
 * File:	modules/Snapper.ycp
 * Summary:	Snapper settings, input and output functions
 * Authors:	Jiri Suchomel <jsuchome@suse.cz>
 *
 * $Id$
 *
 * Representation of the configuration of snapper.
 * Input and output routines.
 */

{

module "Snapper";
textdomain "snapper";

import "FileUtils";
import "Label";
import "Progress";
import "Report";


// global list of all snapshot
global list<map> snapshots	= [];

global string snapshots_path	= "/snapshots";

global map selected_snapshot	= $[];

// mapping of snapshot number to index in snapshots list
global map<integer,integer> id2index	= $[];

// index to snapshots list
global integer selected_snapshot_index	= 0;

/**
 * Return map of files modified between given snapshots
 * Return structure has just one level, and maps each modified file to it's modification map
 */
global map<string,string> ReadModifiedFilesIndex (integer from, integer to) {
    return (map<string,string>) SCR::Read (.snapper.diff_index, $[ "from" : from, "to" : to]);
}

/**
 * Return map of files modified between given snapshots
 * Map is recursively describing the filesystem structure; helps to build Tree widget contents
 */
global map<string, map> ReadModifiedFilesMap (integer from, integer to) {
    return (map<string, map>) SCR::Read (.snapper.diff_tree, $[ "from" : from, "to" : to]);
}

/**
 * Describe what was done with given file between given snapshots
 * - when new is 0, meaning is 'current system'
 */
global map GetFileModification (string file, integer old, integer new) {

    map ret	= $[];

    string file1	= sformat ("%1/%2/snapshot%3", snapshots_path, old, file);
    string file2	= sformat ("%1/%2/snapshot%3", snapshots_path, new, file);
    if (new == 0)
	file2	= file;

    y2milestone ("comparing '%1' and '%2'", file1, file2);

    if (FileUtils::Exists (file1) && FileUtils::Exists (file2))
    {
	list<string> status	= [ "no_change" ];
	map out	= (map) SCR::Execute (.target.bash_output,
	    sformat ("/usr/bin/diff -u %1 %2", file1, file2));
	if (out["stderr"]:"" != "")
	{
	    y2warning ("out: %1", out);
	    ret["diff"] = out["stderr"]:"";
	}
	// the file diff
	else if (out["stdout"]:nil != "")
	{
	    status	= ["diff"];
	    ret["diff"]	= out["stdout"]:"";
	}

	// check mode and ownerships
	out = (map) SCR::Execute (.target.bash_output,
	    sformat ("ls -ld %1 %2 | cut -f 1,3,4 -d ' '", file1, file2));
	list<string> parts	= splitstring (out["stdout"]:""," \n");

	if (parts[0]:"" != parts[3]:"")
	{
	    status		= add (status, "mode");
	    ret["mode1"]	= parts[0]:"";
	    ret["mode2"]	= parts[3]:"";
	}
	if (parts[1]:"" != parts[4]:"")
	{
	    status		= add (status, "user");
	    ret["user1"]	= parts[1]:"";
	    ret["user2"]	= parts[4]:"";
	}
	if (parts[2]:"" != parts[5]:"")
	{
	    status		= add (status, "group");
	    ret["group1"]	= parts[2]:"";
	    ret["group2"]	= parts[5]:"";
	}
	ret["status"]	= status;
    }
    else if (FileUtils::Exists (file1))
    {
	ret["status"]	= ["removed"];
    }
    else if (FileUtils::Exists (file2))
    {
	ret["status"]	= ["created"];
    }
    else
    {
	ret["status"]	= ["none"];
    }
    return ret;
}

/**
 * Read the list of snapshots
 */
global boolean ReadSnapshots () {


    list<map> snapshot_maps	= (list<map>) SCR::Read (.snapper.snapshots);
    if (snapshot_maps == nil) snapshot_maps = [];
    integer i	= 0;
    foreach (map snapshot, snapshot_maps, {
	integer id	= snapshot["num"]:0;
	if (id == 0) return; // ignore the 'current system'
	snapshot["name"]	= tostring (id);
	y2debug ("snapshot data: %1", snapshot);
	snapshots	= add (snapshots, snapshot);
	id2index[id]	= i;
	i		= i + 1;
    });


    return true;
}

global map LastSnapperErrorMap () {

    return (map) SCR::Read (.snapper.error);
}

/**
 * Initialize snapper agent
 * Return true on success
 */
global boolean InitializeSnapper () {

    boolean init	= (boolean) SCR::Execute (.snapper);
    if (!init)
    {
	map err_map	= LastSnapperErrorMap ();
	string type	= err_map["type"]:"";
	string details	= _("Reason not known.");
	if (type == "config_not_found")
	{
	    details	= _("Configuration not found.");
	}
	else if (type == "config_invalid")
	{
	    details	= _("Configuration is not valid.");
	}

	y2warning ("init failed with '%1'", err_map);
	// error popup
	Report::Error (sformat (_("Failed to initialize snapper library:
%1"), details));
    }
    return init;

}

/**
 * Read all snapper settings
 * @return true on success
 */
global boolean Read() {

    /* Snapper read dialog caption */
    string caption = _("Initializing Snapper");

    integer steps = 2;

    // We do not set help text here, because it was set outside
    Progress::New( caption, " ", steps, [
	    /* Progress stage 1/3 */
	    _("Read the list of snapshots"),
	], [
	    /* Progress step 1/3 */
	    _("Reading the database..."),
	    /* Progress finished */
	    _("Finished")
	],
	""
    );

    Progress::NextStage();

    if (!InitializeSnapper ())
    {
	return false;
    }

    ReadSnapshots ();

    Progress::NextStage();
    return true;
}

/**
 * Return the given file mode as octal number
 */
integer GetFileMode (string file) {

    map out = (map) SCR::Execute (.target.bash_output, "/bin/stat --printf=%a " + file);
    return tointeger (out["stdout"]:"755");
}

/**
 * Copy given files from selected snapshot to current filesystem
 * @param snapshot_num snapshot identifier
 * @param files list of full paths to files
 * @return success
 */
global boolean RestoreFiles (integer snapshot_num, list<string> files) {

    boolean ret	= true;
    y2milestone ("going to restore files %1", files);

    UI::OpenDialog (`opt (`decorated), `HBox (`HSpacing(1.5), `VBox (
	`HSpacing (60),
	// label for log window
	`LogView (`id (`log), _("Restoring Files..."), 8, 0),
	`ProgressBar (`id (`progress), "", size (files), 0),
	`PushButton (`id (`ok), Label::OKButton ())
    ), `HSpacing (1.5)));

    UI::ChangeWidget (`id (`ok), `Enabled, false);
    integer progress	= 0;
    foreach (string file, files, {

	UI::ChangeWidget (`id (`progress ), `Value, progress);

	string orig	= sformat ("%1/%2/snapshot", snapshots_path, snapshot_num) + file;

	string dir	= substring (file, 0, findlastof (file, "/"));
	if (!FileUtils::Exists (orig))
	{
	    // log entry (%1 is file name)
	    UI::ChangeWidget (`id (`log), `LastLine, sformat (_("%1 does not exist in snapshot %2\n"), file, snapshot_num));
	}
	else if (FileUtils::CheckAndCreatePath (dir))
	{
	    y2milestone ("copying '%1' to '%2' (dir: %3)", orig, file, dir);
	    if (FileUtils::IsDirectory (orig) == true)
	    {
		map stat	= (map) SCR::Read (.target.stat, orig);
		if (!FileUtils::Exists (file))
		{
		    SCR::Execute (.target.mkdir, file);
		}
		SCR::Execute (.target.bash, sformat ("/bin/chown %1:%2 '%3'", stat["uid"]:0, stat["gid"]:0, file));
		SCR::Execute (.target.bash, sformat ("/bin/chmod %1 '%2'", GetFileMode (orig), file));
	    }
	    else
	    {
		SCR::Execute (.target.bash, sformat ("/bin/cp -a '%1' '%2'", orig, file));
	    }
	    UI::ChangeWidget (`id (`log), `LastLine, file + "\n");
	}
	else
	{
	    y2milestone ("failed to copy file '%1' to '%2' (dir: %3)", orig, file, dir);
	    // log entry (%1 is file name)
	    UI::ChangeWidget (`id (`log), `LastLine, sformat (_("%1 skipped\n"), file));
	}
	sleep (100);
	progress        = progress + 1;
    });

    UI::ChangeWidget (`id (`progress ), `Value, progress);
    UI::ChangeWidget (`id (`ok), `Enabled, true);

    UI::UserInput ();
    UI::CloseDialog ();

    return ret;
}


/* EOF */
}
