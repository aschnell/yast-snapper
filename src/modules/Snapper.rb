# encoding: utf-8

# ------------------------------------------------------------------------------
# Copyright (c) 2006-2012 Novell, Inc. All Rights Reserved.
#
#
# This program is free software; you can redistribute it and/or modify it under
# the terms of version 2 of the GNU General Public License as published by the
# Free Software Foundation.
#
# This program is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
# FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License along with
# this program; if not, contact Novell, Inc.
#
# To contact Novell about this file by physical or electronic mail, you may find
# current contact information at www.novell.com.
# ------------------------------------------------------------------------------

# File:	modules/Snapper.ycp
# Summary:	Snapper settings, input and output functions
# Authors:	Jiri Suchomel <jsuchome@suse.cz>
#
# $Id$
#
# Representation of the configuration of snapper.
# Input and output routines.
require "yast"

module Yast
  class SnapperClass < Module
    def main
      Yast.import "UI"
      textdomain "snapper"

      Yast.import "FileUtils"
      Yast.import "Label"
      Yast.import "Progress"
      Yast.import "Report"
      Yast.import "String"

      # global list of all snapshot
      @snapshots = []

      @selected_snapshot = {}

      # mapping of snapshot number to index in snapshots list
      @id2index = {}

      # index to snapshots list
      @selected_snapshot_index = 0

      # list of configurations
      @configs = []

      @current_config = ""

    end

    # Return map of files modified between given snapshots
    # Return structure has just one level, and maps each modified file to it's modification map
    def ReadModifiedFilesIndex(from, to)
      Convert.convert(
        SCR.Read(path(".snapper.diff_index"), { "from" => from, "to" => to }),
        :from => "any",
        :to   => "map <string, map>"
      )
    end

    # Return map of files modified between given snapshots
    # Map is recursively describing the filesystem structure; helps to build Tree widget contents
    def ReadModifiedFilesMap(from, to)
      Convert.convert(
        SCR.Read(path(".snapper.diff_tree"), { "from" => from, "to" => to }),
        :from => "any",
        :to   => "map <string, map>"
      )
    end

    # Return the path to given snapshot
    def GetSnapshotPath(snapshot_num)
      ret = Convert.to_string(
        SCR.Read(path(".snapper.path"), { "num" => snapshot_num })
      )
      if ret == nil
        ret = ""
        # popup error
        Report.Error(
          Builtins.sformat(_("Snapshot '%1' was not found."), snapshot_num)
        )
      end
      ret
    end

    # Return the full path to the given file from currently selected configuration (subvolume)
    # @param [String] file path, relatively to current config
    # GetFileFullPath ("/testfile.txt") -> /abc/testfile.txt for /abc subvolume
    def GetFileFullPath(file)
      Ops.get_string(
        @snapshots,
        [@selected_snapshot_index, "files_index", file, "full_path"],
        file
      )
    end

    # Describe what was done with given file between given snapshots
    # - when new is 0, meaning is 'current system'
    def GetFileModification(file, old, new)
      ret = {}
      file1 = Builtins.sformat("%1%2", GetSnapshotPath(old), file)
      file2 = Builtins.sformat("%1%2", GetSnapshotPath(new), file)
      file2 = GetFileFullPath(file) if new == 0

      Builtins.y2milestone("comparing '%1' and '%2'", file1, file2)

      if FileUtils.Exists(file1) && FileUtils.Exists(file2)
        status = ["no_change"]
        out = Convert.to_map(
          SCR.Execute(
            path(".target.bash_output"),
            Builtins.sformat(
              "/usr/bin/diff -u '%1' '%2'",
              String.Quote(file1),
              String.Quote(file2)
            )
          )
        )
        if Ops.get_string(out, "stderr", "") != ""
          Builtins.y2warning("out: %1", out)
          Ops.set(ret, "diff", Ops.get_string(out, "stderr", ""))
        # the file diff
        elsif Ops.get(out, "stdout") != ""
          status = ["diff"]
          ret["diff"] = out["stdout"].encode(Encoding::UTF_8, { :invalid => :replace })
        end

        # check mode and ownerships
        out = Convert.to_map(
          SCR.Execute(
            path(".target.bash_output"),
            Builtins.sformat(
              "ls -ld -- '%1' '%2' | cut -f 1,3,4 -d ' '",
              String.Quote(file1),
              String.Quote(file2)
            )
          )
        )
        parts = Builtins.splitstring(Ops.get_string(out, "stdout", ""), " \n")

        if Ops.get(parts, 0, "") != Ops.get(parts, 3, "")
          status = Builtins.add(status, "mode")
          Ops.set(ret, "mode1", Ops.get(parts, 0, ""))
          Ops.set(ret, "mode2", Ops.get(parts, 3, ""))
        end
        if Ops.get(parts, 1, "") != Ops.get(parts, 4, "")
          status = Builtins.add(status, "user")
          Ops.set(ret, "user1", Ops.get(parts, 1, ""))
          Ops.set(ret, "user2", Ops.get(parts, 4, ""))
        end
        if Ops.get(parts, 2, "") != Ops.get(parts, 5, "")
          status = Builtins.add(status, "group")
          Ops.set(ret, "group1", Ops.get(parts, 2, ""))
          Ops.set(ret, "group2", Ops.get(parts, 5, ""))
        end
        Ops.set(ret, "status", status)
      elsif FileUtils.Exists(file1)
        Ops.set(ret, "status", ["removed"])
      elsif FileUtils.Exists(file2)
        Ops.set(ret, "status", ["created"])
      else
        Ops.set(ret, "status", ["none"])
      end
      deep_copy(ret)
    end

    # Read the list of snapshots
    def ReadSnapshots
      @snapshots = []
      snapshot_maps = Convert.convert(
        SCR.Read(path(".snapper.snapshots")),
        :from => "any",
        :to   => "list <map>"
      )
      snapshot_maps = [] if snapshot_maps == nil
      i = 0
      Builtins.foreach(snapshot_maps) do |snapshot|
        id = Ops.get_integer(snapshot, "num", 0)
        next if id == 0 # ignore the 'current system'
        Ops.set(snapshot, "name", Builtins.tostring(id))
        Builtins.y2debug("snapshot data: %1", snapshot)
        @snapshots = Builtins.add(@snapshots, snapshot)
        Ops.set(@id2index, id, i)
        i = Ops.add(i, 1)
      end
      true
    end


    def LastSnapperErrorMap
      Convert.to_map(SCR.Read(path(".snapper.error")))
    end


    def ReadConfigs
      @configs = Convert.convert(
        SCR.Read(path(".snapper.configs")),
        :from => "any",
        :to   => "list <string>"
      )

      if @configs == nil || @configs.empty?
        return false
      end

      if @configs.include?("root")
        @current_config = "root"
      else
        @current_config = @configs[0]
      end

      return true
    end


    # Initialize snapper agent
    # Return true on success
    def InitializeSnapper(config)
      init = Convert.to_boolean(
        SCR.Execute(path(".snapper"), { "config" => config })
      )
      if !init
        err_map = LastSnapperErrorMap()
        type = Ops.get_string(err_map, "type", "")
        details = _("Reason not known.")
        if type == "config_not_found"
          details = _("Configuration not found.")
        elsif type == "config_invalid"
          details = _("Configuration is not valid.")
        end

        Builtins.y2warning("init failed with '%1'", err_map)
        # error popup
        Report.Error(
          Builtins.sformat(
            _("Failed to initialize snapper library:\n%1"),
            details
          )
        )
      end
      return init
    end


    # Delete existing snapshot
    # Return true on success
    def DeleteSnapshot(args)
      args = deep_copy(args)
      success = Convert.to_boolean(SCR.Execute(path(".snapper.delete"), args))
      if !success
        err_map = LastSnapperErrorMap()
        type = Ops.get_string(err_map, "type", "")
        details = _("Reason not known.")

        details = _("Snapshot was not found.") if type == "not_found"

        Builtins.y2warning("deleting failed with '%1'", err_map)
        # error popup
        Report.Error(
          Builtins.sformat(_("Failed to delete snapshot:\n%1"), details)
        )
      end
      success
    end
    # Modify existing snapshot
    # Return true on success
    def ModifySnapshot(args)
      args = deep_copy(args)
      success = Convert.to_boolean(SCR.Execute(path(".snapper.modify"), args))
      if !success
        err_map = LastSnapperErrorMap()
        type = Ops.get_string(err_map, "type", "")
        details = _("Reason not known.")

        Builtins.y2warning("modification failed with '%1'", err_map)
        # error popup
        Report.Error(
          Builtins.sformat(_("Failed to modify snapshot:\n%1"), details)
        )
      end
      success
    end

    # Create new snapshot
    # Return true on success
    def CreateSnapshot(args)
      args = deep_copy(args)
      success = Convert.to_boolean(SCR.Execute(path(".snapper.create"), args))
      if !success
        err_map = LastSnapperErrorMap()
        type = Ops.get_string(err_map, "type", "")
        details = _("Reason not known.")

        if type == "wrong_snapshot_type"
          details = _("Wrong snapshot type given.")
        elsif type == "pre_not_given"
          details = _("'Pre' snapshot was not given.")
        elsif type == "pre_not_found"
          details = _("Given 'Pre' snapshot was not found.")
        end

        Builtins.y2warning("creating failed with '%1'", err_map)
        # error popup
        Report.Error(
          Builtins.sformat(_("Failed to create new snapshot:\n%1"), details)
        )
      end
      success
    end

    # Read all snapper settings
    # @return true on success
    def Read
      # Snapper read dialog caption
      caption = _("Initializing Snapper")

      steps = 2

      # We do not set help text here, because it was set outside
      Progress.New(
        caption,
        " ",
        steps,
        [
          # Progress stage 1/3
          _("Read the list of snapshots")
        ],
        [
          # Progress step 1/3
          _("Reading the database..."),
          # Progress finished
          _("Finished")
        ],
        ""
      )

      Progress.NextStage

      if !ReadConfigs()
        # error popup
        Report.Error(_("No snapper configurations exist. You have to create one or more
configurations to use yast2-snapper. The snapper command line
tool can be used to create configurations."))
        return false
      end

      return false if !InitializeSnapper(@current_config)

      ReadSnapshots()

      Progress.NextStage
      true
    end

    # Return the given file mode as octal number
    def GetFileMode(file)
      out = Convert.to_map(
        SCR.Execute(
          path(".target.bash_output"),
          Builtins.sformat("/bin/stat --printf=%%a '%1'", String.Quote(file))
        )
      )
      mode = Ops.get_string(out, "stdout", "")
      return 644 if mode == nil || mode == ""
      Builtins.tointeger(mode)
    end

    # Copy given files from selected snapshot to current filesystem
    # @param [Fixnum] snapshot_num snapshot identifier
    # @param [Array<String>] files list of full paths to files
    # @return success
    def RestoreFiles(snapshot_num, files)
      files = deep_copy(files)
      ret = true
      Builtins.y2milestone("going to restore files %1", files)

      UI.OpenDialog(
        Opt(:decorated),
        HBox(
          HSpacing(1.5),
          VBox(
            HSpacing(60),
            # label for log window
            LogView(Id(:log), _("Restoring Files..."), 8, 0),
            ProgressBar(Id(:progress), "", Builtins.size(files), 0),
            PushButton(Id(:ok), Label.OKButton)
          ),
          HSpacing(1.5)
        )
      )

      UI.ChangeWidget(Id(:ok), :Enabled, false)
      progress = 0
      Builtins.foreach(files) do |file|
        UI.ChangeWidget(Id(:progress), :Value, progress)
        orig = Ops.add(GetSnapshotPath(snapshot_num), file)
        full_path = GetFileFullPath(file)
        dir = Builtins.substring(
          full_path,
          0,
          Builtins.findlastof(full_path, "/")
        )
        if !FileUtils.Exists(orig)
          SCR.Execute(
            path(".target.bash"),
            Builtins.sformat("/bin/rm -rf -- '%1'", String.Quote(full_path))
          )
          Builtins.y2milestone("removing '%1' from system", full_path)
          # log entry (%1 is file name)
          UI.ChangeWidget(
            Id(:log),
            :LastLine,
            Builtins.sformat(_("Deleted %1\n"), full_path)
          )
        elsif FileUtils.CheckAndCreatePath(dir)
          Builtins.y2milestone(
            "copying '%1' to '%2' (dir: %3)",
            orig,
            file,
            dir
          )
          if FileUtils.IsDirectory(orig) == true
            stat = Convert.to_map(SCR.Read(path(".target.stat"), orig))
            if !FileUtils.Exists(full_path)
              SCR.Execute(path(".target.mkdir"), full_path)
            end
            SCR.Execute(
              path(".target.bash"),
              Builtins.sformat(
                "/bin/chown -- %1:%2 '%3'",
                Ops.get_integer(stat, "uid", 0),
                Ops.get_integer(stat, "gid", 0),
                String.Quote(full_path)
              )
            )
            SCR.Execute(
              path(".target.bash"),
              Builtins.sformat(
                "/bin/chmod -- %1 '%2'",
                GetFileMode(orig),
                String.Quote(full_path)
              )
            )
          else
            SCR.Execute(
              path(".target.bash"),
              Builtins.sformat(
                "/bin/cp -a -- '%1' '%2'",
                String.Quote(orig),
                String.Quote(full_path)
              )
            )
          end
          UI.ChangeWidget(Id(:log), :LastLine, Ops.add(full_path, "\n"))
        else
          Builtins.y2milestone(
            "failed to copy file '%1' to '%2' (dir: %3)",
            orig,
            full_path,
            dir
          )
          # log entry (%1 is file name)
          UI.ChangeWidget(
            Id(:log),
            :LastLine,
            Builtins.sformat(_("%1 skipped\n"), full_path)
          )
        end
        Builtins.sleep(100)
        progress = Ops.add(progress, 1)
      end

      UI.ChangeWidget(Id(:progress), :Value, progress)
      UI.ChangeWidget(Id(:ok), :Enabled, true)

      UI.UserInput
      UI.CloseDialog

      ret
    end

    publish :variable => :snapshots, :type => "list <map>"
    publish :variable => :selected_snapshot, :type => "map"
    publish :variable => :id2index, :type => "map <integer, integer>"
    publish :variable => :selected_snapshot_index, :type => "integer"
    publish :variable => :configs, :type => "list <string>"
    publish :variable => :current_config, :type => "string"
    publish :function => :ReadModifiedFilesIndex, :type => "map <string, map> (integer, integer)"
    publish :function => :ReadModifiedFilesMap, :type => "map <string, map> (integer, integer)"
    publish :function => :GetSnapshotPath, :type => "string (integer)"
    publish :function => :GetFileFullPath, :type => "string (string)"
    publish :function => :GetFileModification, :type => "map (string, integer, integer)"
    publish :function => :ReadSnapshots, :type => "boolean ()"
    publish :function => :LastSnapperErrorMap, :type => "map ()"
    publish :function => :ReadConfigs, :type => "boolean ()"
    publish :function => :InitializeSnapper, :type => "boolean (string)"
    publish :function => :DeleteSnapshot, :type => "boolean (map)"
    publish :function => :ModifySnapshot, :type => "boolean (map)"
    publish :function => :CreateSnapshot, :type => "boolean (map)"
    publish :function => :Read, :type => "boolean ()"
    publish :function => :RestoreFiles, :type => "boolean (integer, list <string>)"
  end

  Snapper = SnapperClass.new
  Snapper.main
end
