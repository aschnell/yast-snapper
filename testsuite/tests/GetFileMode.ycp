{
    // testedfiles: Snapper.ycp

    import "Snapper";
    import "Testsuite";

    map EX = $[
	"target"	: $[
	    "bash_output"	: $[
		"stdout"	: "755"
	    ],
	]
    ];
    Testsuite::Test (``(Snapper::GetFileMode ("/tmp/1")), [ $[], $[], EX ], 0);

    EX["target","bash_output"]	= $[
	"stdout"	: "",
	"stderr"	: "/bin/stat: cannot stat `/tmp/2': No such file or directory\n"
    ];

    Testsuite::Test (``(Snapper::GetFileMode ("/tmp/2")), [ $[], $[], EX ], 0);

}
