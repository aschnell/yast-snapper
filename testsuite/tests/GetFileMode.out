Execute	.target.bash_output "/bin/stat --printf=%a '/tmp/1'" $["stdout":"755"]
Return	755
Execute	.target.bash_output "/bin/stat --printf=%a '/tmp/2'" $["stderr":"/bin/stat: cannot stat `/tmp/2': No such file or directory\n", "stdout":""]
Return	644
